An elimination constraint graph to manage elimination of sort variables

jrosain opened the pull request:
This PR adds the `QGraph`, a new view of the acyclic graph that manages elimination. It allows actual elimination between sort variables, and even though this PR does not introduce the syntax to so, it is used in multiple places (as a generalization of the previous system or for so-called "above prop" sort variables).

SkySkimmer commented on lib/acyclicGraph.ml:
```diff
@@ -184,8 +187,8 @@ module Make (Point:Point) = struct
     with Not_found ->
-      CErrors.anomaly ~label:"Univ.repr"
-        Pp.(str"Universe " ++ Point.raw_pr u ++ str" undefined.")
+      CErrors.anomaly ~label:Point.anomaly_label
+        @@ Point.anomaly_err u
```
```suggestion
      CErrors.anomaly ~label:Point.anomaly_label (Point.anomaly_err u)
```
I think older ocaml doesn't like combining labeled arguments with @@

jrosain replied:
Oh, thanks for pointing it out, I wouldn't have found it.

SkySkimmer commented:
They're probably the same bug so let's minimize one and see if fixing it also fixes the others
@coqbot ci minimize ci-ext_lib

jrosain commented:
> They're probably the same bug so let's minimize one and see if fixing it also fixes the others @coqbot ci minimize ci-ext_lib

It seems like it is. Should be good now. I might have to add an overlay for metarocq, but I'm not 100% sure.

SkySkimmer commented:
@coqbot run full ci

SkySkimmer commented:
@coqbot ci minimize

SkySkimmer commented:
@JasonGross that doesn't look minimized

SkySkimmer commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,170 @@
+    | QType, _ -> true
+    | QProp, (QProp | QSProp) -> true
+    | QSProp, QSProp -> true
+    | _ -> false
```
```suggestion
    | (QType | QProp | QSProp), _ -> false
```
that way ocaml will tell us this needs to be changed if we add another constant quality

SkySkimmer commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,170 @@
+    | QConstant QType, _ -> true
+    | QConstant q, QConstant q' -> const_eliminates_to q q'
+    | QVar q, QVar q' -> QVar.equal q q'
+    | _ -> false
```
```suggestion
    | (QConstant _ | QVar _), _ -> false
```

SkySkimmer commented on engine/uState.ml:
```diff
@@ -142,6 +155,19 @@ let nf_quality m = function
+  (* Here, we filter instead of enforcing equality due to the collapse:
+     simply enforcing equality may lead to inconsistencies after it *)
+  let qs = QVar.Set.filter filter qs in
+  let fold v g = try QGraph.add_quality (QVar v) g with _ -> g in
```
do not use catchall
I guess this should be
```suggestion
  let fold v g = try QGraph.add_quality (QVar v) g with QGraph.AlreadyDeclared -> g in
```

jrosain commented:
After investigation of the errors triggered by compiling `equations`, I've found out that they come from calls with an empty evar map on the plugin side. I'll speak with @mattam82 to understand how to fix that properly.

jrosain commented:
@coqbot run full ci

SkySkimmer commented:
Also maybe worth squashing the commits

SkySkimmer commented:
@coqbot bench

SkySkimmer commented on kernel/environ.mli:
```diff
@@ -74,9 +74,11 @@ val named_context : env -> Constr.named_context
 
-val qualities : env -> Sorts.QVar.Set.t
-val set_qualities : Sorts.QVar.Set.t -> env -> env
+val qualities     : env -> QGraph.t
+val qvars         : env -> Sorts.QVar.Set.t
```
```suggestion
val qualities : env -> QGraph.t
val qvars : env -> Sorts.QVar.Set.t
```

SkySkimmer commented on kernel/sorts.ml:
```diff
@@ -279,7 +269,7 @@ module QConstraint = struct
       else Quality.compare b b'
 
-  let trivial (a,(Equal|Leq),b) = Quality.equal a b
+  let trivial (a,_,b) = Quality.equal a b
```
it's also possible that all uses of `trivial` are now incorrect and should be using a qgraph, in which case delete `trivial` (cf my comment on conversion.ml)

jrosain commented on kernel/cClosure.mli:
```diff
@@ -129,6 +129,7 @@ type evar_handler = {
   evar_repack : Evar.t * constr list -> constr;
   evar_irrelevant : constr pexistential -> bool;
   qvar_irrelevant : Sorts.QVar.t -> bool;
+  elim_to : Sorts.Quality.t -> Sorts.Quality.t -> bool;
```
I'm not planning on adding a use of elim_to. It impacts a bit more than these two functions though (c.f. the recent push).

SkySkimmer commented on kernel/qGraph.mli:
```diff
@@ -33,22 +33,40 @@ type explanation =
+type constraint_source =
+  | Internal
+  | Rigid
+  | Static
```
how is this meant to work? currently AFAICT only Internal is used, so what is supposed to be using Rigid / Static?

jrosain replied:
Static should be used when wanting to skip all tests (except the one already done in `enforce_*` in the acyclic graph). It's already used in `add_quality` to enforce that Type eliminates to every sort.

Rigid will be for a constraint entered by the user, e.g.:
```coq
Sort s.
Constraint s -> SProp. (* Rigid *)

Definition foo@{s'|...| s -> s'} ... (* also rigid *)
```

SkySkimmer replied:
What happens if the user enters (rigid) `Prop -> s` and `s -> Type` (implying `Prop -> Type`)?

SkySkimmer commented:
We may also want to forbid `SProp -> s` (for `s <> SProp`), I think it would cause trouble with definitional proof irrelevance otherwise.

jrosain commented:
> We may also want to forbid `SProp -> s` (for `s <> SProp`), I think it would cause trouble with definitional proof irrelevance otherwise.

I think that one may want to declare such a constraint. But it seems sensible to forbid it now and enable it in the future when we're sure that it doesn't break anything.

SkySkimmer commented:
looks unrelated, I see it failing in https://github.com/rocq-prover/rocq/pull/20634 too

jrosain commented:
@coqbot run full ci

SkySkimmer commented:
perf looks not great on some fiat crypto commands

jrosain commented:
> perf looks not great on some fiat crypto commands

Added a cache in the ustate for qualities eliminating to prop. Performances might be better with this.

jrosain commented:
> commit " Removed now-useless above_prop from UGraph and reworked check_[l]eq_sort" should be removed instead of adding back above_prop afterwards

You're right. I'll be squashing the commits next, I wanted to check whether I introduced a bug in the full CI or not before that.

jrosain commented:
> @jrosain did you forget to tell coqbot to run the bench?

Ah, it seems that the bench did not launch. Is it the right syntax to run the bench with the overlays, and did I simply forget to mention coqbot?

SkySkimmer commented:
There is no syntax to run the bench with overlays, you have to pass them to https://github.com/rocq-prover/rocq/blob/f4e25061958980d0fc28144cd3b40587ba6c7724/dev/bench/bench.sh#L55-L57 (by modifying the script)
To run the bench you then do coqbot bench (with the `@`)

mattam82 commented:
@coqbot run full ci

jrosain commented:
@coqbot bench

jrosain commented:
@coqbot bench

jrosain commented:
@coqbot run full ci

SkySkimmer commented:
For elpi probably because it ran before https://github.com/rocq-prover/opam/pull/3466 got merged
For equations maybe you need to rebase the overlay (overlays get auto rebased in CI, but in the bench it just uses the branch you give it)

jrosain commented:
Thanks!

jrosain commented:
@coqbot bench

jrosain commented:
@coqbot run full ci

SkySkimmer commented on kernel/environ.mli:
```diff
@@ -386,10 +387,10 @@ val push_context_set : ?strict:bool -> ContextSet.t -> env -> env
-(** [push_qualities qs env] pushes the set of quality variables in
-    the environment. It fails if a quality variable is already
-    declared. *)
+val push_quality_set : Sorts.QVar.Set.t -> env -> env
```
rebase went wrong? this probably shouldn't have changed (note that the comment is wrong, the current version of the PR does fail is a qvar is already declared)

SkySkimmer reviewed (commented):
seems almost ready modulo the comment
(also remove the bench overlay commit)

jrosain commented:
@coqbot run full ci

jrosain commented:
> Unsurprisingly the slowdown is due to an overhead where we keep unifying qualities for what seems to be freshly generated terms in rewrite. This is just making worse a code that's already dubious at best. I'm not sure there is much to do unfortunately...

I see, thanks for looking into it At least we know for the future that there is a room for improvement on this front.

jrosain commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,330 @@
+  enforce_constraint Internal (s1, ElimConstraint.Equal, s2) g
+
+let initial_graph() =
+  match !cached_graph with
```
Do you mean in the call to `enforce_constraint`?
If so, this is about creating an `ElimConstraint`. I could use a constructor `make_elim_constraint s1 ElimConstraint.Equal s2` instead?

ppedrot replied:
No, I'm talking about the deferred computation with `cached_graph`. I don't understand why you have to do it this way, especially given that it'll be immediately computed in Environ. Just get rid of the reference and compute the result once and for all.

jrosain replied:
Yes, that's a relic of a time I needed to add template qvars in the graph. It's fixed now, thanks.

ppedrot commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,330 @@
+  str "Cannot enforce" ++ spc() ++ Quality.pr prv q1 ++ spc() ++
+    ElimConstraint.pr_kind k ++ spc() ++ Quality.pr prv q2 ++ spc() ++ reason
+
+let explain_elimination_error defprv err =
```
This function (and probably bits of the one above) should be moved to Himsg.

ppedrot commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,330 @@
+  match to_graph_cstr k with
+  | Le -> G.check_leq
+  | Eq -> G.check_eq
+  | Lt -> assert false
```
This assertion is useless, you could match directly on `k` instead of adding dead code.

ppedrot commented on kernel/qGraph.ml:
```diff
@@ -0,0 +1,330 @@
+  match to_graph_cstr k with
+  | Le -> G.enforce_leq
+  | Eq -> G.enforce_eq
+  | Lt -> assert false
```
Same remark as above.

ppedrot commented on kernel/qGraph.mli:
```diff
@@ -0,0 +1,105 @@
+  | Other of Pp.t
+
+type quality_inconsistency =
+  ((QVar.t -> Pp.t) option) * (ElimConstraint.kind * Quality.t * Quality.t * explanation option)
```
AFAICT the `((QVar.t -> Pp.t) option)` value is fortunately only set to `None`.

ppedrot replied:
Just remove it for now, we'll discuss the details when actually implementing this.

jrosain replied:
Sure thing, done!
